torch/nn/functional.py

```diff
@@ -3510,6 +3510,170 @@ def cross_entropy(
     )
 
 
+def _linear_cross_entropy_naive(
+    input: Tensor,
+    linear_weight: Tensor,
+    target: Tensor,
+    linear_bias: Optional[Tensor],
+    reduction: str,
+    ignore_index: int,
+    label_smoothing: float,
+) -> Tensor:
+    logits = linear(input, linear_weight, linear_bias)
+    logits_flat = logits.reshape(-1, logits.size(-1))
+    target_flat = target.reshape(-1)
+    loss = cross_entropy(
+        logits_flat,
+        target_flat,
+        reduction=reduction,
+        ignore_index=ignore_index,
+        label_smoothing=label_smoothing,
+    )
+    if reduction == "none":
+        loss = loss.reshape(target.shape)
+    return loss
+
+
+def linear_cross_entropy(
+    input: Tensor,
+    linear_weight: Tensor,
+    target: Tensor,
+    *,
+    linear_bias: Optional[Tensor] = None,
+    reduction: str = "mean",
+    ignore_index: int = -100,
+    label_smoothing: float = 0.0,
+    chunking_strategy: str = "none",
+    vocab_chunk_size: Optional[int] = None,
+    batch_chunk_size: Optional[int] = None,
+) -> Tensor:
+    r"""Compute fused linear transformation and cross entropy loss on CPU.
+
+    This is a convenience wrapper around :func:`linear` followed by
+    :func:`cross_entropy`.  When the inputs live on CPU it uses a fused ATen
+    kernel that chunks the vocabulary or batch dimension to avoid materialising
+    large logit tensors; the chunk sizes can be overridden with
+    ``vocab_chunk_size`` and ``batch_chunk_size``.  For other devices it falls
+    back to the unfused composition.
+    """
+    if has_torch_function_variadic(input, linear_weight, target, linear_bias):
+        return handle_torch_function(
+            linear_cross_entropy,
+            (input, linear_weight, target, linear_bias),
+            input,
+            linear_weight,
+            target,
+            linear_bias=linear_bias,
+            reduction=reduction,
+            ignore_index=ignore_index,
+            label_smoothing=label_smoothing,
+            chunking_strategy=chunking_strategy,
+            vocab_chunk_size=vocab_chunk_size,
+            batch_chunk_size=batch_chunk_size,
+        )
+
+    if not isinstance(reduction, str):
+        if hasattr(reduction, "node"):
```

Comment:
why is this needed whereas other functions in this file that take in string arguments (e.g. `embedding_bag`) don't have this sort of logic